utils/s2n_map.c

```diff
@@ -246,3 +246,77 @@ S2N_RESULT s2n_map_free(struct s2n_map *map)
 
     return S2N_RESULT_OK;
 }
+
+S2N_RESULT s2n_map_size(struct s2n_map *map, uint32_t *size)
+{
+    RESULT_ENSURE_REF(map);
+    *size = map->size;
+    return S2N_RESULT_OK;
+}
+
+/* Update the internal state so that `current_index` will point to the next value
+ * in the table or set iter->consumed equal to true if there are no more elements
+ * in the map. 
+ */
+S2N_RESULT s2n_map_iterator_advance(struct s2n_map_iterator *iter)
+{
+    RESULT_ENSURE_REF(iter);
+    RESULT_ENSURE_REF(iter->map);
+    RESULT_ENSURE(!iter->consumed, S2N_ERR_SAFETY);
+
+    iter->current_index++;
+    while (iter->current_index < iter->map->capacity) {
```

Comment:
Not in a particularly neat way. Choose the greater evil, `while (true)` or two increments 😄 . The bounds check has to occur after the index increment, which restricts the choices that we have. I tend to avoid `while(true)` loops since I like the condition check to be semantically valid, but happy to change it if you have a different preference.